addons/popochiu/engine/objects/region/popochiu_region.gd

```diff
@@ -121,48 +126,71 @@ func _check_scaling(
 	area_rid: RID, area: Area2D, area_shape_index: int, local_shape_index: int, entered: bool
 ):
 	if not (
-		area is PopochiuCharacter 
-		and area.get_node_or_null("ScalingPolygon") 
-		and area_shape_index == area.get_node("ScalingPolygon").get_index()
+		area is PopochiuCharacter
+		and area.get("scaling_polygon")
+		and area_shape_index == area.get("scaling_polygon").get_index()
 	):
 		return
 	
-	if not entered:
-		_clear_scaling_region(area)
-		return
+	var character: PopochiuCharacter = area
+	# Track character entry/exit across all shapes
+	if entered:
+		_active_characters[character.script_name] = area
+	else:
+		# Only remove if character is not in any other shape
+		var overlapping := get_overlapping_areas()
+		var still_inside := false
+		for overlapping_area in overlapping:
+			if overlapping_area == character:
```

Comment:
I understand what's going on here but I think the naming of the variables are very hard to get.

Comparing an overlapping area with a character makes little sense. Can't we come up with a different naming? Or better yet, can't this whole loop be implemented using find_key() or get()?

Reply:
Let's see if the new approach feels better.